File: ros2_ws/src/vision_node/vision_node/protocol.py
"""
[레거시 — 현재 사용 안 함]
초기 개발 단계 <CMD,VALUE,CS> XOR 체크섬 프로토콜 헬퍼임. 이건 아두이노
펌웨어 시리얼 통신 프로토타입 시절 규약이고, 지금 vision_node.py는
/vision/detected를 JSON 문자열로 발행함(json.dumps). 이 프로토콜은 안 씀.
dummy_subscriber.py가 이걸 참조하는데, dummy_subscriber도 같이 레거시임.
"""
"""<CMD,VALUE,CS> 프로토콜 헬퍼 (펌웨어 protocol.cpp와 동일 규약).

vision_node가 /vision/detected로 내보내는 문자열과 dummy_subscriber의 검증이
동일한 XOR 체크섬 규약을 공유하도록 한곳에 모은다.
"""
from __future__ import annotations


def xor_checksum(payload: str) -> int:
    """payload 문자열의 모든 문자를 XOR 한 1바이트(0~255) 체크섬."""
    cs = 0
    for ch in payload:
        cs ^= ord(ch)
    return cs & 0xFF


def build_message(cmd: str, value: str) -> str:
    """'<CMD,VALUE,CS>' 형식 문자열 생성 (CS = 2자리 대문자 HEX)."""
    payload = f"{cmd},{value}"
    return f"<{payload},{xor_checksum(payload):02X}>"


def parse_message(line: str) -> tuple[str, str] | None:
    """'<CMD,VALUE,CS>' 파싱 + 체크섬 검증.

    성공하면 (cmd, value)를, 형식/체크섬 오류면 None을 반환한다.
    """
    line = line.strip()
    if not line.startswith("<") or not line.endswith(">"):
        return None
    body = line[1:-1]
    parts = body.rsplit(",", 1)
    if len(parts) != 2:
        return None
    payload, cs_str = parts
    try:
        got = int(cs_str, 16)
    except ValueError:
        return None
    if got != xor_checksum(payload):
        return None
    cmd, _, value = payload.partition(",")
    return cmd, value

File: archive/firmware_esp32_arm/arm_firmware/protocol.cpp
/*
 * protocol.cpp - <CMD,VALUE,CS> 프로토콜 구현
 */
#include "protocol.h"

// payload 문자열의 각 문자를 누적 XOR -> 1바이트 체크섬
uint8_t xorChecksum(const char* payload) {
  uint8_t cs = 0;
  for (const char* p = payload; *p; ++p) {
    cs ^= (uint8_t)(*p);
  }
  return cs;
}

// "CMD" + "," + "VALUE" 를 만들고 체크섬을 붙여 한 줄 메시지로 반환
String buildMessage(const char* cmd, const char* value) {
  String payload = String(cmd) + "," + value;        // "CMD,VALUE"
  uint8_t cs = xorChecksum(payload.c_str());
  char hex[3];
  snprintf(hex, sizeof(hex), "%02X", cs);            // 2자리 대문자 HEX
  return String("<") + payload + "," + hex + ">\n";  // "<CMD,VALUE,CS>\n"
}

// CMD/VALUE 문자열을 내부 CommandType으로 매핑
static CommandType mapCommand(const String& cmd, const String& val, int& outValue) {
  outValue = 0;
  if (cmd == "POSE") {
    if (val == "HOME")  return CMD_POSE_HOME;
    if (val == "READY") return CMD_POSE_READY;
    if (val == "DROP")  return CMD_POSE_DROP;
  } else if (cmd == "WRIST") {
    outValue = val.toInt();   // 방위각 0~180
    return CMD_WRIST;
  } else if (cmd == "GRIP") {
    if (val == "SOFT") return CMD_GRIP_SOFT;
    if (val == "HARD") return CMD_GRIP_HARD;
    if (val == "OPEN") return CMD_GRIP_OPEN;
  } else if (cmd == "STOP") {
    return CMD_STOP;
  }
  return CMD_UNKNOWN;
}

// "<CMD,VALUE,CS>" 한 줄을 파싱. 프레임/콤마/체크섬을 모두 검증한다.
bool parseMessage(const String& line, ArmCommand& out) {
  int lt = line.indexOf('<');
  int gt = line.indexOf('>');
  if (lt < 0 || gt < 0 || gt <= lt) return false;     // 프레임(<...>) 검증
  String body = line.substring(lt + 1, gt);           // "CMD,VALUE,CS"

  int c1 = body.indexOf(',');       // 첫 콤마 (CMD 끝)
  int c2 = body.lastIndexOf(',');   // 마지막 콤마 (CS 앞)
  if (c1 < 0 || c2 <= c1) return false;               // 콤마 2개 필요

  String cmd   = body.substring(0, c1);
  String val   = body.substring(c1 + 1, c2);
  String csStr = body.substring(c2 + 1);

  // 체크섬 재계산 후 비교
  String payload = cmd + "," + val;
  uint8_t expect = xorChecksum(payload.c_str());
  uint8_t got = (uint8_t)strtol(csStr.c_str(), nullptr, 16);
  if (expect != got) return false;                    // 체크섬 불일치 -> 폐기

  out.type = mapCommand(cmd, val, out.value);
  return out.type != CMD_UNKNOWN;
}

const char* armStateName(ArmState s) {
  switch (s) {
    case ARM_IDLE:    return "IDLE";
    case ARM_MOVING:  return "MOVING";
    case ARM_GRIPPED: return "GRIPPED";
    default:          return "ERROR";
  }
}
